peptonizer_js/src/py-worker.js

```diff
@@ -1,27 +1,91 @@
 
 const callbacks = {};
+let id = 0; // identify a Promise
 
-pyodideWorker.onmessage = (event) => {
+// Called if graph generation is done.
+generatePepgmGraphWorker.onmessage = (event) => {
     const { id, ...data } = event.data;
     const onSuccess = callbacks[id];
     delete callbacks[id];
     onSuccess(data);
 };
 
-const asyncRun = (() => {
-    let id = 0; // identify a Promise
+// Start generating the graph.
+const pepgmGraphGeneration = (() => {
     return (context) => {
         // the id could be generated more carefully
         id = (id + 1) % Number.MAX_SAFE_INTEGER;
         return new Promise((onSuccess) => {
             callbacks[id] = onSuccess;
-            pyodideWorker.postMessage({
+            // Start a worker to generate the graph.
+            generatePepgmGraphWorker.postMessage({
                 ...context,
                 id,
             });
         });
     };
 })();
 
-export { asyncRun };
\ No newline at end of file
+// Keep one pepgm result at each point of execution.
+// Once a new result is ready, compare it to the current result and choose the best solution.
+let result;
+const aggregateResult = (taskResult) => {
+    result = taskResult;
+};
+
+// Create numWorkers workers, which will execute pepGM for different parameters in parallel.
+const numWorkers = 2
+const executePepgmWorkers = []
+let tasksDone = 0;
+for (let i = 0; i < numWorkers; i++) {
+    let worker = new Worker("./workers/execute_pepgm_worker.js");
+    // Define what to do if a pepGM execution is done.
+    worker.onmessage = (event) => {
+        const { id, ...data } = event.data;
+        const onSuccess = callbacks[id];
+        delete callbacks[id];
+        
+        // Aggregate the results
+        aggregateResult(data);
+
+        // call onSuccess if all parameter combinations are done.
+        tasksDone ++;
+        if (tasksDone == tasksAmount) {
+            onSuccess(result);
+        }
+    };
+    executePepgmWorkers.push(worker);
+}
+
+// Execute pepGM with all given parameter combinations.
+let tasksAmount = 0;
+const asyncPepgmExecution = (() => {
+    return (context) => {
+        const { graph, alphas, betas, priors } = context;
+
+        return new Promise((onSuccess) => {
+            // Loop through all parameter combinations.
+            alphas.forEach((alpha, i, as) => {
+                betas.forEach((beta, j, bs) => {
+                    priors.forEach((prior, k, ps) => {
+                        id = (id + 1) % Number.MAX_SAFE_INTEGER;
+                        callbacks[id] = onSuccess;
+                        // Distribute the tasks among the workers.
+                        executePepgmWorkers[tasksAmount % numWorkers].postMessage({
+                            graph,
+                            id,
+                            alpha,
+                            beta,
+                            prior
+                        });
```

Comment:
You're right here that messages send to a worker will be queued (if a new message arrives to the worker before the previous one is finished). You should, however, be careful in doing so because it is not guaranteed that the worker fully finishes the task before starting a new one. This will not mess up our results, but it might result in a higher memory usage. See this thread for more information: https://stackoverflow.com/questions/28608254/if-i-send-multiple-messages-to-the-same-webworker-does-it-queue-them-up-and-pro

You can take a look at the "async" library, and more specifically the `queue` function. This allows you to build a worker pool to which you can send tasks to and execute them in parallel using _n_ workers at the same time: https://caolan.github.io/async/v3/docs.html#queue

For more information on how to do this, you can take a look at Unipept (I've already implemented something similar there): https://github.com/unipept/unipept-web-components/blob/876ea2f108d99da47c76b33e008e2583997f5259/src/logic/util/queue/WorkerQueue.ts

Reply:
I implemented this request, together with the previous two. Can you re-review the new parts of code?